File: src/main/java/ru/xpendence/topjavagraduation/repository/VoteRepository.java
package ru.xpendence.topjavagraduation.repository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import ru.xpendence.topjavagraduation.entity.Vote;

import java.time.LocalDate;
import java.util.Optional;

public interface VoteRepository extends JpaRepository<Vote, Long> {

    @Query("SELECT v FROM Vote v JOIN FETCH v.user JOIN FETCH v.restaurant WHERE v.id = :id")
    Optional<Vote> findByIdWithDetails(@Param("id") Long id);

    @Query("SELECT v FROM Vote v JOIN FETCH v.user JOIN FETCH v.restaurant WHERE v.user.id = :userId AND v.date = :date")
    Optional<Vote> findByUserIdAndDate(@Param("userId") Long userId, @Param("date") LocalDate date);
}

File: src/main/kotlin/ru/xpendence/topjavagraduation/repository/VoteRepository.kt
package ru.xpendence.topjavagraduation.repository

import ru.xpendence.topjavagraduation.entity.Vote
import java.time.LocalDate
import java.util.*

/**
 * Доступ к голосам через JDBC.
 */
interface VoteRepository {

    /**
     * Сохраняет голос: вставляет новую запись или обновляет дату и ресторан существующей по [Vote.id].
     *
     * @param vote голос; для вставки — без id, с user.id и restaurant.id
     * @return сохранённый голос (с присвоенным id при вставке)
     */
    fun save(vote: Vote): Vote

    /**
     * Возвращает голос по идентификатору без полной загрузки связанных сущностей.
     *
     * @param id идентификатор голоса
     * @return голос или empty, если не найден
     */
    fun findById(id: Long): Optional<Vote>

    /**
     * Возвращает голос по идентификатору с загруженными пользователем и рестораном.
     *
     * @param id идентификатор голоса
     * @return голос с деталями или empty, если не найден
     */
    fun findByIdWithDetails(id: Long): Optional<Vote>

    /**
     * Возвращает голос пользователя на указанную дату с загруженными пользователем и рестораном.
     *
     * @param userId идентификатор пользователя
     * @param date дата голосования
     * @return голос или empty, если не найден
     */
    fun findByUserIdAndDate(userId: Long, date: LocalDate): Optional<Vote>
}
